GoF creational patterns

AntonEsta opened the pull request:
Паттерны GoF: Порождающие.

Задание.
1. Применить один из порождающих паттернов для приложения “Картотека домашних животных”
2. Написать собственную реализацию паттерна Singleton.

Singleton реализован в классе CardFile.

AntonEsta committed:
Revert "Add package "test". Add class "test.main". Add method "test.test.main""

This reverts commit bb527878

AntonEsta committed:
Homework #04. Create package "lec04" with "task.txt"

AntonEsta committed:
Homework #04. Added package "lec04.task01".
Added classes: Card, CardFile, Person, Pet, prog.
Removed file task.txt to lec04.task01

AntonEsta committed:
Homework #04. Refactoring. Remove file Card.java

AntonEsta committed:
Homework #04. Refactoring use Project Lombok java library.

AntonEsta committed:
Homework #11. Pet card file with using creational patterns. Master commit.

efremov-itcenter commented on src/main/java/petscardfile/classes/Person.java:
```diff
@@ -0,0 +1,27 @@
+    private int age;
+    @NonNull private Sex sex;
+
+    public Person(PersonBuilder builder) {
```
Ох, билдер лучше прямо в здесь описать, как статический класс, зачем его выносить отдельным классом?

AntonEsta replied:
+ Класс PersonBuilder перенесен в класс Person.

efremov-itcenter commented on src/main/java/petscardfile/classes/Person.java:
```diff
@@ -0,0 +1,27 @@
+@ToString
+@EqualsAndHashCode
+public class Person {
+    @NonNull private String name;
```
Тогда уж поля можно сделать финальными

AntonEsta replied:
+ Поля 'name', 'surname' и 'sex' объявлен как final.

efremov-itcenter commented on src/main/java/petscardfile/classes/Pet.java:
```diff
@@ -0,0 +1,25 @@
+    @Override
+    public String toString() {
+        Formatter ft = new Formatter();
+        ft.format("nickname: %6s; weight: %3.1f; owner: [%s] %n", nickname, weight, owner);
```
`String.format(...)`

AntonEsta replied:
+ Метод Pet.toString() перписан с использованием String.format(...).

efremov-itcenter commented on src/main/java/petscardfile/classes/card/CardFile.java:
```diff
@@ -0,0 +1,114 @@
+     * @param nickname кличка питомца.
+     * @return {@link CardFile} найденных питомцев или {@code null} если негде искать.
+     */
+    public HashMap<@NonNull UUID, Pet> findPet(String nickname){
```
Лучше в возвращаемом значении не привязываться к конкретной реализации, возвращайте `Map<UUID, Pet>`

AntonEsta replied:
+ Возвращаемое значение переопределено.

efremov-itcenter commented on src/main/java/petscardfile/classes/card/CardFile.java:
```diff
@@ -0,0 +1,114 @@
+     * раньше выводит животное, у которого больше вес.
+     * @return список животных в отсортированном порядке.
+     */
+    public ArrayList<Pet> sortedPetList(){
```
Тоже самое, просто `List<Pet>`

AntonEsta replied:
+ Возвращаемое значение переопределено.

efremov-itcenter commented on src/main/java/petscardfile/classes/factories/AbstractFactory.java:
```diff
@@ -0,0 +1,25 @@
+import petscardfile.classes.card.CardFile;
+import petscardfile.classes.enums.Sex;
+
+public class AbstractFactory {
```
Глядя на имя, ожидаешь, что AbstractFactory будет действительно абстрактным классом, но нет, здесь вполне конкретная реализация

AntonEsta replied:
+ Класс переименован на ObjectFactory.

efremov-itcenter commented on src/main/java/petscardfile/classes/card/CardFile.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    public HashMap<@NonNull UUID, Pet> findPet(String nickname){
+        if (pets.isEmpty()) return null;
+        Map<@NonNull UUID, Pet> map = pets.entrySet().stream()
+                .filter(e -> e.getValue().getNickname().equals(nickname))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```
Я бы похоливарил, что здесь поиск по имени можно считать эффективным, хотя конечно, "эффективность" - понятие очень растяжимое, и в определенных условиях линейную сложность можно считать достаточно эффективной.

AntonEsta committed:
Homework #11. Revision of the code according to the comments of pull request.

+ Static class PersonBuilder was moved into the Person class with name Builder.

+ Privet fields 'name', 'surname' and 'sex' of Person class were defined as finals.

+ Method Pet.toString() was implemented with String.format(String format, Object... args) method.

+ Method CardFile.findPet. Was been redefined the return type from HashMap to Map.

+ + Method CardFile.sortedPetList. Was been redefined the return type from ArrayList to List.

+ Class AbstractFactory was renamed to ObjectFactory.
